lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1464,34 +1460,66 @@ static bool validateTBDIfNeeded(const CompilerInvocation &Invocation,
                            Opts, allSymbols);
 }
 
+enum class DeallocatableResources {
+  None,
+  SILModule,
+  SILModuleAndASTContext,
+};
+static DeallocatableResources
+canFreeUpCompilerResources(const CompilerInvocation &Invocation,
+                           CompilerInstance &Instance) {
+  // If the stats reporter is installed, we need the ASTContext and SILModule
+  // to live through the entire compilation process.
+  if (Instance.getASTContext().Stats) {
+    return DeallocatableResources::None;
+  }
+
+  // If we're going to dump the API of the module, we cannot tear down
+  // the ASTContext, as that would cause the module to be freed prematurely.
+  if (!Invocation.getFrontendOptions().DumpAPIPath.empty()) {
+    return DeallocatableResources::SILModule;
+  }
+
+  // If there are multiple primary inputs it is too soon to free
+  // the ASTContext, etc.. OTOH, if this compilation generates code for > 1
+  // primary input, then freeing it after processing the last primary is
+  // unlikely to reduce the peak heap size. So, only optimize the
+  // single-primary-case (or WMO).
+  if (Invocation.getFrontendOptions()
+          .InputsAndOutputs.hasMultiplePrimaryInputs()) {
+    return DeallocatableResources::SILModule;
+  }
+
+  return DeallocatableResources::SILModuleAndASTContext;
```

Comment:
Very nicely documented. I wish there were some way to centralize the invariants. 
The fact that the stats option requires the `ASTContext` and `SILModule`, that the dump API path option requires the `ASTContext`, etc., ought to be somewhere else, but I have no good ideas.